File: DesktopCore/src/util/UpdateChecker.java
package util;

import java.io.IOException;

import javax.swing.JOptionPane;

import org.w3c.dom.Document;
import org.w3c.dom.Element;

/**
 * Class to check for updates and launch the EntropyUpdater via a batch file if they are available
 */
public class UpdateChecker implements XmlConstants
{
	public static void checkForUpdates(String filename, int portForDownload)
	{
		try
		{
			checkForUpdatesAndDoDownloadIfRequired(filename, portForDownload);
		}
		finally
		{
			DialogUtil.dismissLoadingDialog();
		}
	}
	
	private static void checkForUpdatesAndDoDownloadIfRequired(String filename, int portForDownload)
	{
		//Show this here, checking the CRC can take time
		DialogUtil.showLoadingDialog("Checking for updates...");
		String crc = FileUtil.getMd5Crc(filename);
		if (crc == null)
		{
			DialogUtil.showError("Failed to check for updates (couldn't find " + filename + ").");
			return;
		}
		
		Debug.append("Checking for updates - fileCrc is " + crc);
		
		//We have a CRC, so go to the Server and check against it's copy.
		Document crcCheck = factoryCrcCheck(crc, filename);
		String responseStr = AbstractClient.getInstance().sendSync(crcCheck, false);
		if (responseStr == null)
		{
			DialogUtil.showError("Failed to check for updates (unable to connect).");
			return;
		}
		
		DialogUtil.dismissLoadingDialog();
		
		Document xmlResponse = XmlUtil.getDocumentFromXmlString(responseStr);
		Element rootElement = xmlResponse.getDocumentElement();
		String responseName = rootElement.getNodeName();
		if (responseName.equals(RESPONSE_TAG_NO_UPDATES))
		{
			//No need to show a message, should be pretty obvious
			Debug.append("I am up to date");
			return;
		}
		
		//An update is available
		int answer = DialogUtil.showQuestion("An update is available. Would you like to download it now?", false);
		if (answer == JOptionPane.NO_OPTION)
		{
			return;
		}
		
		startUpdate(rootElement, filename, portForDownload);
	}
	
	private static Document factoryCrcCheck(String fileCrc, String fileName)
	{
		Document document = XmlUtil.factoryNewDocument();
		Element rootElement = document.createElement(ROOT_TAG_CRC_CHECK);
		rootElement.setAttribute("FileCrc", fileCrc);
		rootElement.setAttribute("FileName", fileName);
		
		document.appendChild(rootElement);
		return document;
	}
	
	public static void startUpdate(Element rootElement, String filename, int portForUpdate)
	{
		int fileSize = XmlUtil.getAttributeInt(rootElement, "FileSize");
		String version = rootElement.getAttribute("VersionNumber");
		String args = fileSize + " " + version + " " + filename + " " + portForUpdate;
		
		try
		{
			if (AbstractClient.isAppleOs())
			{
				Runtime.getRuntime().exec("update.command " + args);
			}
			else
			{
				Runtime.getRuntime().exec("cmd /c start update.bat " + args);
			}
		}
		catch (IOException ioe)
		{
			Debug.stackTrace(ioe);
			String manualCommand = "update.bat " + args;
			
			String msg = "Failed to launch update.bat - call the following manually to perform the update: \n\n" + manualCommand;
			DialogUtil.showError(msg);
		}
		
		System.exit(0);
	}
}

File: Darts/src/burlton/dartzee/code/utils/UpdateChecker.kt
package burlton.dartzee.code.utils

import burlton.core.code.util.Debug
import burlton.desktopcore.code.util.DialogUtil
import com.mashape.unirest.http.Unirest
import org.json.JSONObject
import java.io.File
import java.io.IOException
import javax.swing.JOptionPane

/**
 * Class to check for updates and launch the EntropyUpdater via a batch file if they are available
 */
object UpdateChecker
{
    fun checkForUpdates()
    {
        try
        {
            checkForUpdatesAndDoDownloadIfRequired()
        }
        finally
        {
            DialogUtil.dismissLoadingDialog()
        }
    }

    private fun checkForUpdatesAndDoDownloadIfRequired()
    {
        //Show this here, checking the CRC can take time
        DialogUtil.showLoadingDialog("Checking for updates...")

        Debug.append("Checking for updates - my version is $DARTS_VERSION_NUMBER")

        val jsonObject = queryLatestReleaseJson()
        if (jsonObject == null)
        {
            DialogUtil.showError("Failed to check for updates (unable to connect).")
            return
        }

        val remoteVersion = jsonObject.get("tag_name").toString()
        if (remoteVersion == DARTS_VERSION_NUMBER)
        {
            Debug.append("I am up to date")
            return
        }

        DialogUtil.dismissLoadingDialog()

        Debug.append("Newer release available - $remoteVersion")
        val assets = jsonObject.getJSONArray("assets")
        if (assets.length() != 1)
        {
            Debug.append(jsonObject.toString())
            Debug.stackTrace("Unexpected number of assets ${assets.length()} - aborting update")
            return
        }

        //An update is available
        val answer = DialogUtil.showQuestion("An update is available ($remoteVersion). Would you like to download it now?", false)
        if (answer == JOptionPane.NO_OPTION)
        {
            return
        }

        startUpdate(remoteVersion, assets.getJSONObject(0))
    }

    private fun queryLatestReleaseJson(): JSONObject?
    {
        try
        {
            val response = Unirest.get("https://api.github.com/repos/alexburlton/Dartzee/releases/latest").asJson()
            if (response.status != 200)
            {
                Debug.append("Received unexpected HTTP response. Status ${response.status} - ${response.statusText}")
                Debug.append(response.body.toString())
                DialogUtil.showError("Failed to check for updates (unable to connect).")
                return null
            }

            return response.body.`object`
        }
        catch (t: Throwable)
        {
            Debug.stackTraceSilently(t)
            return null
        }
    }

    private fun startUpdate(remoteVersion: String, asset: JSONObject)
    {
        val assetId = asset.getLong("id")
        val fileName = asset.getString("name")
        val size = asset.getLong("size")

        val updateFile = File("update.bat")
        if (!updateFile.exists())
        {
            //Write the batch file now
            val updateScript = javaClass.getResource("/update/update.bat").readText()
            updateFile.writeText(updateScript)
        }

        val args = "$size $remoteVersion $fileName $assetId"

        try
        {
            Runtime.getRuntime().exec("cmd /c start update.bat $args")
            System.exit(0)
        }
        catch (ioe: IOException)
        {
            Debug.stackTrace(ioe)
            val manualCommand = "update.bat $args"

            val msg = "Failed to launch update.bat - call the following manually to perform the update: \n\n$manualCommand"
            DialogUtil.showError(msg)
        }
    }
}
